### PR TITLE
[CUDA] Fill out push constant kernel arguments at dispatch time

### PR DESCRIPTION
**Pull request by ThomasRaoux**
This is needed as push constant command may not be sent again

### CODE REVIEW

File: iree/hal/cuda/graph_command_buffer.c [hunk 1]
```diff
@@ -18,6 +18,11 @@
 #include "iree/hal/cuda/native_executable.h"
 #include "iree/hal/cuda/status_util.h"
 
+#define IREE_HAL_CUDA_MAX_PUSH_CONSTANT 64
```
Comment (benvanik) [inline iree/hal/cuda/graph_command_buffer.c:21]:
```suggestion
#define IREE_HAL_CUDA_MAX_PUSH_CONSTANT_COUNT 64
```

File: iree/hal/cuda/graph_command_buffer.c [hunk 2]
```diff
@@ -425,7 +425,17 @@ static iree_status_t iree_hal_cuda_graph_command_buffer_dispatch(
...
       iree_hal_cuda_graph_command_buffer_cast(base_command_buffer);
-
+  iree_hal_executable_layout_t* layout =
+      iree_hal_cuda_executable_get_layout(executable, entry_point);
+  iree_host_size_t num_constants =
+      iree_hal_cuda_executable_layout_num_constants(layout);
+  iree_host_size_t constant_base_index =
+      iree_hal_cuda_push_constant_index(layout);
+  // Patch the push constants in the kernel arguments.
+  for (iree_host_size_t i = 0; i < num_constants; i++) {
```
Comment (benvanik) [inline iree/hal/cuda/graph_command_buffer.c:435]: do you validate the max num_constants is under the limit when creating the layout? (should match IREE_HAL_CUDA_MAX_PUSH_CONSTANT_COUNT, so you may want to move that to your executable_layout.h so you can share it in both)

File: iree/hal/cuda/native_executable.c [hunk 3]
```diff
@@ -86,6 +92,7 @@ iree_status_t iree_hal_cuda_native_executable_create(
     executable->entry_functions[i].block_size_x = block_sizes_vec[i].x;
     executable->entry_functions[i].block_size_y = block_sizes_vec[i].y;
     executable->entry_functions[i].block_size_z = block_sizes_vec[i].z;
+    executable->executable_layouts[i] = executable_spec->executable_layouts[i];
```
Comment (benvanik) [inline iree/hal/cuda/native_executable.c:95]: you'll want to retain the layout on the executable and have the executable release it when done; see
https://github.com/google/iree/blob/2f7888bc4491e578ae73637613820be012ac8a3b/iree/hal/local/local_executable.c#L23 and https://github.com/google/iree/blob/2f7888bc4491e578ae73637613820be012ac8a3b/iree/hal/local/local_executable.c#L40

**Review by benvanik — CHANGES_REQUESTED** [review]

File: iree/hal/cuda/graph_command_buffer.c (hunk 1 above)
Comment (ThomasRaoux) [reply iree/hal/cuda/graph_command_buffer.c]: Done

File: iree/hal/cuda/graph_command_buffer.c (hunk 2 above)
Comment (ThomasRaoux) [reply iree/hal/cuda/graph_command_buffer.c:434]: right, added that

File: iree/hal/cuda/native_executable.c (hunk 3 above)
Comment (ThomasRaoux) [reply iree/hal/cuda/native_executable.c:95]: thanks, added those here and in the destroy function

**Review by benvanik — APPROVED** [review]

**Pull request merged**